sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -239,8 +246,53 @@ private TestPipeline(
   public Statement apply(final Statement statement, final Description description) {
     return new Statement() {
 
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation annotation) {
+          return annotation.annotationType() != null
+              && annotation.annotationType().equals(Category.class);
+        }
+      };
+
+      private final Predicate<Annotation> categoryOfNeedsRunner = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation category) {
+          return
+              FluentIterable
+                  .from(Arrays.asList(((Category) category).value()))
+                  .anyMatch(new Predicate<Class<?>>() {
+
+                    @Override
+                    public boolean apply(final Class<?> aClass) {
+                      return NeedsRunner.class.isAssignableFrom(aClass);
+                    }
+                  });
+        }
+      };
+
+      private void setDeducedEnforcementLevel() {
+        if (enforcement instanceof PipelineAbandonedNodeEnforcement) {
+
+          final boolean annotatedWithNeedsRunner =
+              FluentIterable
+                  .from(description.getAnnotations())
+                  .filter(hasCategoryAnnotation)
+                  .anyMatch(categoryOfNeedsRunner);
+
+          final boolean crashingRunner =
+              CrashingRunner.class.isAssignableFrom(getOptions().getRunner());
+
+          enableAbandonedNodeEnforcement(annotatedWithNeedsRunner || !crashingRunner);
```

Comment:
If a test is annotated with `NeedsRunner` and has a `CrashingRunner`, there's a configuration problem in whatever is running the tests. It might be worth having a `checkState` here to validate that.